framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1429,63 +1443,74 @@ protected void checkTypecastSafety(TypeCastTree node, Void p) {
+                    && newExprType.getKind() == TypeKind.DECLARED) {
+                int castSize = ((AnnotatedDeclaredType) newCastType).getTypeArguments().size();
+                int exprSize = ((AnnotatedDeclaredType) newExprType).getTypeArguments().size();
+
+                if (castSize != exprSize) {
+                    // Always warn if the cast and expression contain a different number of
+                    // type arguments, e.g. to catch a cast from "Object" to "List<@NonNull Object>".
+                    // TODO: the same number of arguments actually doesn't guarantee anything.
+                    return false;
                 }
+            } else if (castType.getKind() == TypeKind.TYPEVAR
+                    && exprType.getKind() == TypeKind.TYPEVAR) {
+                // If both the cast type and the casted expression are type variables, then check the
+                // bounds.
+                Set<AnnotationMirror> lowerBoundAnnotationsCast =
+                        AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                                qualifierHierarchy, castType);
+                Set<AnnotationMirror> lowerBoundAnnotationsExpr =
+                        AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                                qualifierHierarchy, exprType);
+                return qualifierHierarchy.isSubtype(
+                                lowerBoundAnnotationsExpr, lowerBoundAnnotationsCast)
+                        && qualifierHierarchy.isSubtype(
+                                exprType.getEffectiveAnnotations(),
+                                castType.getEffectiveAnnotations());
+            }
+            Set<AnnotationMirror> castAnnos;
+            if (castType.getKind() == TypeKind.TYPEVAR) {
+                // If the cast type is a type var, but the expression is not, then check that the
+                // type of the expression is a subtype of the lower bound.
+                castAnnos =
+                        AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                                qualifierHierarchy, castType);
             } else {
-                // Only check the main qualifiers, ignoring array components and
-                // type arguments.
-                isSubtype =
-                        atypeFactory
-                                .getQualifierHierarchy()
-                                .isSubtype(
-                                        exprType.getEffectiveAnnotations(),
-                                        castType.getEffectiveAnnotations());
+                castAnnos = castType.getAnnotations();
             }
-        }
 
-        if (!isSubtype) {
-            checker.report(Result.warning("cast.unsafe", exprType, castType), node);
+            return qualifierHierarchy.isSubtype(exprType.getEffectiveAnnotations(), castAnnos);
         }
+        return qualifierHierarchy.isSubtype(
```

Comment:
I would move this into an `else` block, to clarify the control flow a little. The only way to get here is if `checkCastElementType` isn't provided, right? Maybe add that as a comment in the else block.